### DISCUSSION TITLE
SpatialResample vs ResampleToMatch vs Spacing

### DISCUSSION THREAD

**Discussion by hermancollin** [Project-MONAI/MONAI#6248] [Q&A] [answered]
Hi. I'm looking into the resampling transforms to get a common pixel size (e.g. 0.1 um/px) for all my 2D images across training/validation/test splits. I'm confused about the distinction between these 3 transforms:
- [SpatialResample](https://docs.monai.io/en/stable/transforms.html#spatialresample)
- [ResampleToMatch](https://docs.monai.io/en/stable/transforms.html#resampletomatch)
- [Spacing](https://docs.monai.io/en/stable/transforms.html#spacing)
Looking at the source code, it seems the last 2 are related to the `SpatialResample` transform. 

I understand an integral part of these transforms is the `affine` matrix inside the image metadata. How does this work for 2D PNG images, for example? From the [load_medical_images](https://github.com/Project-MONAI/tutorials/blob/main/modules/load_medical_images.ipynb) tutorial, I can see the `LoadImaged` transform will read a PNG image using PILReader and give a 4x4 identity matrix as affine matrix. How can I embed the pixel size of my images in this affine matrix?

**Comment by wyli** [comment, marked as answer]
one possible option would be making a customized transform to assign the pixel size and SpacingD to normalize them:

```py
import os
import numpy as np
import monai
from monai.transforms import Compose, Lambda, LoadImageD, SpacingD


class AssignPixelSize:
    def __init__(self, key, pixel_key):
        self.key = key
        self.pixel_key = pixel_key

    def __call__(self, data_dict):
        print(data_dict[self.pixel_key])
        data_dict[self.key].affine[0, 0] = data_dict[self.pixel_key][0]
        data_dict[self.key].affine[1, 1] = data_dict[self.pixel_key][1]
        return data_dict


filename = os.path.join(".", "MONAI-logo_color.png")
monai.apps.download_url("https://monai.io/assets/img/MONAI-logo_color.png", filepath=filename)
transform = Compose(
    [
        LoadImageD(keys="image", image_only=True, ensure_channel_first=True, dtype=np.uint8),
        Lambda(func=AssignPixelSize(key="image", pixel_key="pixel_size")),
        SpacingD(keys="image", pixdim=(0.1, 0.1)),
    ]
)
test_data = {"image": filename, "pixel_size": (0.18, 0.17)}
result = transform(test_data)
print(f"image data shape:{result['image'].shape}")
print(result["image"].affine)
```

**Reply by hermancollin** [reply]
@wyli thank you! This is exactly what I was looking for. I tested it and it works as expected.
